crates/orderbook/src/order_quoting.rs

```diff
@@ -332,19 +368,202 @@ pub trait OrderQuoteCalculating: Send + Sync {
     async fn find_quote(&self, search: Search) -> Result<Quote, FindQuoteError>;
 }
 
-pub enum CalculateQuoteError {}
+#[derive(Error, Debug)]
+pub enum CalculateQuoteError {
+    #[error("sell amount does not cover fee")]
+    SellAmountDoesNotCoverFee(U256),
+
+    #[error("failed to estimate price")]
+    Price(#[from] PriceEstimationError),
+
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+#[derive(Error, Debug)]
+pub enum FindQuoteError {
+    #[error("quote not found")]
+    NotFound(Option<QuoteId>),
+
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+/// Fields for searching stored quotes.
+pub struct QuoteSearchParameters {
+    pub sell_token: H160,
+    pub buy_token: H160,
+    pub sell_amount: U256,
+    pub buy_amount: U256,
+    pub kind: OrderKind,
+}
+
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait QuoteStoring: Send + Sync {
+    /// Saves a quote and returns its ID.
+    async fn save(&self, data: QuoteData) -> Result<QuoteId>;
+
+    /// Retrieves an existing quote by ID.
+    async fn get(&self, id: QuoteId, expiration: DateTime<Utc>) -> Result<Option<QuoteData>>;
+
+    /// Retrieves an existing quote by ID.
+    async fn find(
+        &self,
+        parameters: QuoteSearchParameters,
+        expiration: DateTime<Utc>,
+    ) -> Result<Option<(QuoteId, QuoteData)>>;
+}
+
+#[cfg_attr(test, mockall::automock)]
+pub trait Now: Send + Sync {
+    fn now(&self) -> DateTime<Utc>;
```

Comment:
Would it be cleaner to work with std Instant and Duration mostly only converting to DateTime at the end?

Reply:
I mostly inherited this from `fee.rs` (which uses `DateTime<Utc>`). 

> cleaner to work with std Instant and Duration mostly only converting to DateTime at the end?


There were no pain-points working with `chrono::DateTime` so I didn't even consider using something else. Can change it if you think their is value.

Reply:
Current way is fine.